### DISCUSSION TITLE
Auditing of Special Events

### DISCUSSION THREAD

**Discussion by ndegwamartin** [opensrp/fhircore#2528] [Ideas]
We need to implement an Audit trail for special actions performed by the practitioners (and other users) on the android client

- [ ] Move to catchment
- [ ] Editing of resources/tasks/action
- [ ] Any changes by _Supervisor_ for data synced to their device via P2P

**Resources**
- https://www.hl7.org/fhir/auditevent.html
- http://hl7.org/fhir/R4B/provenance.html

**Comment by pld** [comment]
Good call, this feels like something to control with a feature flag as we work on it. I wonder if we could implement this with the decorator pattern, maybe using annotations

**Comment by ekigamba** [comment]
If we assumed that the user interacting with the FHIR server API automatically made the change on the frontend device, would we be able to handle this audit logging under this issue https://github.com/onaio/opensrp-tasks/issues/107 ?

Maybe review the issue and the scope to make sure it serves this purpose also

**Reply by pld** [reply]
One exception to that assumption is when we use p2p sync to transfer data from healthcare devices to a manager device and only the manager device syncs w/the remote server

**Reply by ekigamba** [reply]
Nice catch :thinking:

**Reply by ekigamba** [reply]
p2p sync generally introduces a number of edge cases. Currently, we add the practitioner-id in the meta for new resources. This could help resolve this at the API level but the fact that this resource can be edited again on the manager device without the practitioner-id being added means we might have to handle this audit at the front-end device

**Comment by ndegwamartin** [comment]
We can implement the normal interactions using the BALP interceptors @ekigamba mentioned on [the conversation here](https://github.com/onaio/opensrp-tasks/issues/107) as they follow a logging FHIR IG [as documented here](https://profiles.ihe.net/ITI/BALP/). 

Then for any special events we can create custom Audit Events on the device using the [AuditEvent](https://www.hl7.org/fhir/auditevent.html) resource.

**Comment by ndegwamartin** [comment]
However for the P2P case above it appears the most appropriate FHIR resource to use would be the [Provenance](http://hl7.org/fhir/R4B/provenance.html) one @roywanyaga mentioned. 

There seems to be a clear distinction in its application vs the _AuditEvent_ resource where the former is generated by the creating entity of the data (_device_) while the latter by the responding entity which is the _server_ in this case.

**Comment by dubdabasoduba** [comment]
@aurangzaibumer raised an interesting item we could log. 
- The number of resources affected by the fhircore migration scripts. e.g. Migration version 1.0.0 ran and affected 20 Tasks.

**Reply by pld** [reply]
Cool, even logging the UUIDs seems reasonable
